File: day08/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	puzzleInput = "input.txt"
)

const (
	CodeNoop = iota
	CodeAcc
	CodeJmp
)

type (
	Instr struct {
		code int
		arg  int
	}

	Machine struct {
		instrs    []Instr
		ip        int
		acc       int
		loopLimit int
	}
)

func NewMachine(instrs []Instr, loopLimit int) *Machine {
	return &Machine{
		instrs:    instrs,
		ip:        0,
		acc:       0,
		loopLimit: loopLimit,
	}
}

func (m *Machine) Step() (bool, error) {
	if m.ip == len(m.instrs) {
		return true, nil
	}
	if m.ip < 0 || m.ip > len(m.instrs) {
		return false, fmt.Errorf("Ip out of bounds: %d", m.ip)
	}
	instr := m.instrs[m.ip]
	switch instr.code {
	case CodeAcc:
		m.acc += instr.arg
		m.ip++
	case CodeJmp:
		m.ip += instr.arg
	case CodeNoop:
		m.ip++
	}
	return false, nil
}

func (m *Machine) Run() (int, error) {
	runSet := map[int]int{}
	for {
		if _, ok := runSet[m.ip]; !ok {
			runSet[m.ip] = 0
		}
		runSet[m.ip]++
		if runSet[m.ip] > m.loopLimit {
			return m.acc, fmt.Errorf("Looped over %d", m.loopLimit)
		}
		term, err := m.Step()
		if err != nil {
			return m.acc, err
		}
		if term {
			return m.acc, nil
		}
	}
}

func parseInstr(line string) (*Instr, error) {
	fields := strings.Fields(line)
	if len(fields) != 2 {
		return nil, fmt.Errorf("Syntax err")
	}
	code := CodeNoop
	switch fields[0] {
	case "acc":
		code = CodeAcc
	case "jmp":
		code = CodeJmp
	}
	arg, err := strconv.Atoi(fields[1])
	if err != nil {
		return nil, err
	}
	return &Instr{
		code: code,
		arg:  arg,
	}, nil
}

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	instrs := []Instr{}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		instr, err := parseInstr(scanner.Text())
		if err != nil {
			log.Fatal(err)
		}
		instrs = append(instrs, *instr)
	}
	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	m := NewMachine(instrs, 1)
	part1, _ := m.Run()
	fmt.Println("Part 1:", part1)

	for n, i := range instrs {
		if i.code == CodeAcc {
			continue
		}
		swapToJmp := i.code == CodeNoop
		if swapToJmp {
			instrs[n] = Instr{
				code: CodeJmp,
				arg:  i.arg,
			}
		} else {
			instrs[n] = Instr{
				code: CodeNoop,
				arg:  i.arg,
			}
		}
		m := NewMachine(instrs, 1)
		part2, err := m.Run()
		if err == nil {
			fmt.Println("Part 2:", part2)
			return
		}
		if swapToJmp {
			instrs[n] = Instr{
				code: CodeNoop,
				arg:  i.arg,
			}
		} else {
			instrs[n] = Instr{
				code: CodeJmp,
				arg:  i.arg,
			}
		}
	}
}

File: day08/src/main.rs
use std::collections::HashMap;
use std::convert::TryFrom;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

#[derive(Clone, Copy, PartialEq, Eq)]
enum Code {
    Noop,
    Acc,
    Jmp,
}

#[derive(Clone, Copy)]
struct Instr(Code, i32);

impl Instr {
    fn parse(line: &str) -> Result<Instr, Box<dyn std::error::Error>> {
        let fields = line.split_whitespace().collect::<Vec<_>>();
        let code = match fields.get(0).ok_or("Invalid line")? {
            &"nop" => Code::Noop,
            &"acc" => Code::Acc,
            &"jmp" => Code::Jmp,
            bogus => return Err(format!("Invalid code: {}", bogus).into()),
        };
        let arg = fields.get(1).ok_or("Invalid line")?.parse::<i32>()?;
        Ok(Instr(code, arg))
    }
}

struct Machine<'a> {
    instrs: &'a Vec<Instr>,
    ip: i32,
    acc: i32,
}

impl<'a> Machine<'a> {
    fn new(instrs: &'a Vec<Instr>) -> Machine<'a> {
        Machine {
            instrs,
            ip: 0,
            acc: 0,
        }
    }

    fn step(&mut self) -> Result<bool, Box<dyn std::error::Error>> {
        let ip = usize::try_from(self.ip)?;
        if ip == self.instrs.len() {
            return Ok(true);
        }
        let Instr(code, arg) = self
            .instrs
            .get(ip)
            .ok_or(format!("Ip out of bounds: {}", self.ip))?;
        match code {
            Code::Noop => self.ip += 1,
            Code::Acc => {
                self.acc += arg;
                self.ip += 1;
            }
            Code::Jmp => self.ip += arg,
        }
        Ok(false)
    }

    fn run(&mut self, loop_limit: usize) -> Result<(), Box<dyn std::error::Error>> {
        let mut run_set = HashMap::new();
        loop {
            let &mut e = run_set.entry(self.ip).and_modify(|e| *e += 1).or_insert(1);
            if e > loop_limit {
                return Err(format!("Looped over {} at {}", loop_limit, self.ip).into());
            }
            if self.step()? {
                return Ok(());
            }
        }
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut instrs = Vec::new();
    for line in reader.lines() {
        let line = line?;
        instrs.push(Instr::parse(&line)?);
    }

    let mut m = Machine::new(&instrs);
    let _ = m.run(1);
    println!("Part 1: {}", m.acc);

    for i in 0..instrs.len() {
        let Instr(code, arg) = instrs[i];
        if code == Code::Acc {
            continue;
        }
        let swap_to_jmp = code == Code::Noop;
        if swap_to_jmp {
            instrs[i] = Instr(Code::Jmp, arg);
        } else {
            instrs[i] = Instr(Code::Noop, arg);
        }
        let mut m = Machine::new(&instrs);
        if let Ok(_) = m.run(1) {
            println!("Part 2: {}", m.acc);
            break;
        }
        if swap_to_jmp {
            instrs[i] = Instr(Code::Noop, arg);
        } else {
            instrs[i] = Instr(Code::Jmp, arg);
        }
    }

    Ok(())
}
